torch/masked/maskedtensor/core.py

```diff
@@ -0,0 +1,556 @@
+
+    def _set_data_mask(self, data, mask):
+        # This method is regrettably necessary for in-place operations
+
+        self._masked_data = data
+        self._masked_mask = mask
+        self._validate_members()
+
+    def __repr__(self):
+        formatter = "{0:8.4f}"
+        if self.dim() == 0:
+            scalar_data = get_data(self).item()
+            data_formatted = (
+                formatter.format(scalar_data)
+                if isinstance(scalar_data, float)
+                else str(scalar_data)
+            )
+            if not get_mask(self).item():
+                data_formatted = "--"
+            return (
+                "MaskedTensor("
+                + data_formatted
+                + ", "
+                + str(get_mask(self).item())
+                + ")"
+            )
+        s = _masked_tensor_str(get_data(self), get_mask(self), formatter)
+        s = "\n".join("  " + si for si in s.split("\n"))
+        return "MaskedTensor(\n" + s + "\n)"
+
+    # Seems like this needs to be defined before torch_dispatch to work
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
+        kwargs = kwargs or {}
+
+        if func in [torch.Tensor.where, torch.where]:
+            _check_args_kwargs_length(args, kwargs, "__torch_function__, torch.where", len_args=3, len_kwargs=0)
+            return MaskedWhere.apply(*args)
+        if func is torch.Tensor.contiguous:
+            return MaskedContiguous.apply(args[0])
+        if func is torch.Tensor.to_dense:
+            return MaskedToDense.apply(args[0])
+        if func is torch.Tensor.to_sparse:
+            return MaskedToSparse.apply(args[0])
+        if func is torch.Tensor.to_sparse_csr:
+            return MaskedToSparseCsr.apply(args[0])
+        if not all(issubclass(cls, t) for t in types):
+            return NotImplemented
+        with torch._C.DisableTorchFunction():
+            ret = func(*args, **kwargs)
+            if func in get_default_nowrap_functions():
+                return ret
+            else:
+                return torch._tensor._convert(ret, cls)
+
+    @classmethod
+    def unary(cls, fn, data, mask):
+        return MaskedTensor(fn(data), mask)
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs):
+        func = func.overloadpacket
+
+        if len(args) <= 0:
+            raise ValueError("It is required that len(args) > 0 for __torch_dispatch__")
+        if func in [torch.ops.aten.mm, torch.ops.aten.bmm]:
+            _check_args_kwargs_length(args, kwargs, f"__torch_dispatch__, {func}", len_args=2, len_kwargs=0)
+            return cls.matmul(args[0], args[1], func)  # type: ignore[call-arg]
+        # Doesn't work for addmm where the first argument is a Tensor
+        data = get_data(args[0])
+        mask = get_mask(args[0])
+        if func is torch.ops.aten.stride:
+            return None
+        if func is torch.ops.prim.layout:
```

Comment:
high-level comment: do we generally recommend implementing `__torch_dispatch__` as a cascade of `if`s? it seems like there's a latent scaling problem, not just w.r.t. readability (which could be helped by factoring out the implementations) but also the linear lookup time. e.g. why not a map from `torch.ops.aten.*` to python functions? @albanD you probably have thoughts on this

Reply:
There isn't requirement from the `__torch_dispatch__` point of view.
But for this kind of things, using a map from op to impl and definition a function for each would make it easier to read for sure.
In particular, you can make the registration decorator based like is done for refs: https://github.com/pytorch/pytorch/blob/0e8beb7d0d1c10e155d650dcf57a8468b3faa423/torch/_refs/fft.py#L242

Reply:
Thanks! Is this registration pattern recommended to the current approach?

Reply:
It is your choice really :)

Reply:
@bhosmer thanks for the call out - we have some tests later in the stack for some of these functions, so maybe I could tack on a diff at the end to change it to this framework?